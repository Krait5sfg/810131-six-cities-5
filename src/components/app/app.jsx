import React from 'react';
import MainPage from '../main-page/main-page';
import PropTypes from 'prop-types';
import {BrowserRouter, Switch, Route} from 'react-router-dom';
import LoginPage from '../login-page/login-page';
import FavoritePage from '../favorite-page/favorite-page';
import OfferPage from '../offer-page/offer-page';
import {OfferPropTypes} from '../../utils/property-type';
import {PagePath, AuthorizationStatus} from '../../utils/const';
import {connect} from 'react-redux';
import {selectCityOffers} from '../../selector/selector';
import PrivateRoute from '../private-route/private-route';

const App = ({allOffers, offers, city, authorizationStatus}) => {

  const favoriteOffers = allOffers.filter((offer) => offer.isFavorite);

  const onLinkEmailClick = (evt, history) => {
    evt.preventDefault();
    history.push(authorizationStatus === AuthorizationStatus.NO_AUTH
      ? PagePath.LOGIN
      : PagePath.FAVORITE);
  };

  const onFavoriteButtonClick = (evt, history) => {
    evt.preventDefault();
    if (authorizationStatus === AuthorizationStatus.NO_AUTH) {
      history.push(PagePath.LOGIN);
    }
  };

  return (
    <BrowserRouter>
      <Switch>
        <Route exact path={PagePath.MAIN} render={({history}) => (
          <MainPage
            offers={offers}
            city={city}
            onLinkEmailClick={(evt) => onLinkEmailClick(evt, history)}
            onFavoriteButtonClick={(evt) => onFavoriteButtonClick(evt, history)}
          />
        )}>
        </Route>
        <Route exact path={PagePath.LOGIN}>
          <LoginPage city={city} />
        </Route>
        <PrivateRoute
          render={({history}) => {
            return (
              <FavoritePage
                favoriteOffers={favoriteOffers}
                onLinkEmailClick={(evt) => onLinkEmailClick(evt, history)} />
            );
          }}
          path={PagePath.FAVORITE}
          exact />
        <Route exact path={`${PagePath.OFFER}:id`} render={({history, match}) => (
          <OfferPage
            idActiveOffer={+match.params.id}
            onLinkEmailClick={(evt) => onLinkEmailClick(evt, history)}
            onFavoriteButtonClick={(evt) => onFavoriteButtonClick(evt, history)}
          />
        )}>
        </Route>
      </Switch>
    </BrowserRouter >
  );
};

App.propTypes = {
  allOffers: PropTypes.arrayOf(OfferPropTypes).isRequired,
  offers: PropTypes.arrayOf(OfferPropTypes).isRequired,
  city: PropTypes.string.isRequired,
  authorizationStatus: PropTypes.string.isRequired
};

const mapStateToProps = (({DATA, PROCESS, USER}) => ({
  allOffers: DATA.allOffers,
  offers: selectCityOffers({DATA, PROCESS}),
  city: PROCESS.city,
  authorizationStatus: USER.authorizationStatus,
}));

export {App};
export default connect(mapStateToProps)(App);
