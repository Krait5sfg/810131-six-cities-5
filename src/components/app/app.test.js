import React from 'react';
import renderer from 'react-test-renderer';
import {App} from './app';
import {BrowserRouter} from 'react-router-dom';


const offers = [
  {
    city: ``,
    accommodation: {
      bedroomsCount: 0,
      features: [],
      guestsLimit: 0,
      isPremium: true,
      price: 241,
      rating: 4.4,
      title: ``,
      type: ``,
    },
    cityLocation: {latitude: 0, longitude: 0, zoom: 0},
    coordinates: [0, 0],
    description: ``,
    host: {
      avatar: ``,
      isPro: true,
      name: ``,
    },
    id: 1,
    images: [],
    isFavorite: true,
    previewImage: ``,
  },
  {
    city: ``,
    accommodation: {
      bedroomsCount: 0,
      features: [],
      guestsLimit: 0,
      isPremium: true,
      price: 0,
      rating: 0,
      title: ``,
      type: ``,
    },
    cityLocation: {latitude: 0, longitude: 0, zoom: 0},
    coordinates: [0, 0],
    description: ``,
    host: {
      avatar: ``,
      isPro: true,
      name: ``,
    },
    id: 1,
    images: [],
    isFavorite: true,
    previewImage: ``,
  },
];

jest.mock(`../private-route/private-route`, () => `PrivateRoute`);
jest.mock(`../offer-page/offer-page`, () => `OfferPage`);
jest.mock(`../login-page/login-page`, () => `LoginPage`);
jest.mock(`../favorite-page/favorite-page`, () => `FavoritePage`);
jest.mock(`../main-page/main-page`, () => `MainPage`);

it(`App render correctly`, () => {
  const tree = renderer.create(<BrowserRouter><App
    offers={offers}
    city={`Dusseldorf`}
    authorizationStatus={`NO_AUTH`}
  /></BrowserRouter>).toJSON();
  expect(tree).toMatchSnapshot();
});
