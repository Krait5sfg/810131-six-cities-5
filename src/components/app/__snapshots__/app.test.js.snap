// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render correctly 1`] = `
<MainPage
  city="Dusseldorf"
  offers={
    Array [
      Object {
        "accommodation": Object {
          "bedroomsCount": 0,
          "features": Array [],
          "guestsLimit": 0,
          "isPremium": true,
          "price": 0,
          "rating": 0,
          "title": "",
          "type": "",
        },
        "city": "",
        "cityLocation": Object {
          "latitude": 0,
          "longitude": 0,
          "zoom": 0,
        },
        "coordinates": Array [
          0,
          0,
        ],
        "description": "",
        "host": Object {
          "avatar": "",
          "isPro": true,
          "name": "",
        },
        "id": 1,
        "images": Array [],
        "isFavorite": true,
        "previewImage": "",
      },
      Object {
        "accommodation": Object {
          "bedroomsCount": 0,
          "features": Array [],
          "guestsLimit": 0,
          "isPremium": true,
          "price": 0,
          "rating": 0,
          "title": "",
          "type": "",
        },
        "city": "",
        "cityLocation": Object {
          "latitude": 0,
          "longitude": 0,
          "zoom": 0,
        },
        "coordinates": Array [
          0,
          0,
        ],
        "description": "",
        "host": Object {
          "avatar": "",
          "isPro": true,
          "name": "",
        },
        "id": 1,
        "images": Array [],
        "isFavorite": true,
        "previewImage": "",
      },
    ]
  }
  onFavoriteButtonClick={[Function]}
  onLinkEmailClick={[Function]}
/>
`;
